### DISCUSSION TITLE
Fresh for real life projects?

### DISCUSSION THREAD

**Discussion by theArianit** [freshframework/fresh#634] [General]
I recently started with deno, after trying node for a while and having a lot of headache with npm versions, package versions, etc. I am loving deno, it reminds me the old school way of web dev, where you focus on code instead of project configuration.
Then I stumbled on fresh, thanks to fireship.
But, recently someone told me that fresh is just a showcase and isn't really intended for real life application. I also saw a video of Ryan Dahl, remix conf, saying the same, I don't know how old that conf is.
Fresh is still young, v1.0.2, but it looks very promising and I like it. So, what's its status, is it a showcase, did it become production ready or is it intended for real life apps?
Thanks for any input.

**Comment by OldManFroggy** [comment]
It was a showcase, but things have changed a little.

I am currently using for a large production app, and there are some issues
in the way of working, and the fact that the framework is evolving fast
(FRESH, that is).

Biggest issues with Fresh for a lot of new users are around getting various
nodejs/npm module to work (although in a few months that will likely not be
an issue), UI Component frameworks (I spent a fair amount of time here, and
in the end ended up using headless-ui and creating a set of re-usable UI
components of my own), and finally how to share state between islands if it
is needed.

Apart from that, combined with Deno, and Deno Deploy the stack is looking
pretty good, and the community is pretty good.

In summary, Fresh is still early and will take a good year to really
mature, until then expect to work a little to get results you need.

Feel free to reach out with any questions, more than willing to share
information from the trenches.


ps. was early adoptor of remix, and also solid/solidstart (still am to an
extent, bvut have reasons for wanting to work with Fresh on my current
large project.)

**Reply by theArianit** [reply]
thanks for the reply. I really wish that people and companies would focus more on deno. deno is how web dev should be (on my opinion). I come from Asp.Net background and I want to focus more on javascript for my personal projects, since it is one language for almost everything. I am trying to avoid npm as much as possible and deno is the way to go for me.
as for fresh, I like the island architecture it is using, I am glad that they are using preact (still learning the basics of jsx/tsx), and I am informed about some of its missing features. I hope that the deno team and the community will bring it where it should be.

**Reply by theArianit** [reply]
since I come from a different background (asp.net), can fresh be used for the backend, like if I want to use it to build just an API? I am already working on some backend with deno, and I thought maybe it's worth to use fresh for the backend. These full stack javascript frameworks are a bit confusing for me on the backend part.

**Reply by OldManFroggy** [reply]
yes, you can, or you can just do in deno/oak and deploy API to deno deploy which is what I do most of the time

for fresh api endpoints,  you can use a route and just render the response, and deal with data / logic in the handler.

I personally, prefer deno for all APIs and call them from handlers in fresh routes  (which all run serverside, just islands are more pure play clientside). Also, deno -> deno deploy for api's is pretty nice DX

ps. I used to work on ASP.NET back in the day, wrote one of first books back in 2001 or so :)

**Reply by theArianit** [reply]
that's the way I like to do things. Just wanted to hear some other opinions. Deno and oak work very well, did some basic CRUD with edgedb 2.0.

> ps. I used to work on ASP.NET back in the day, wrote one of first books back in 2001 or so :)

that must be 2002, since asp.net was officially released in the beginning of that year, but maybe 2001 since books must be ready when something gets released. awesome

**Reply by OldManFroggy** [reply]
Publish date was november 2001 I was using pre release a year or so before, very early pdc release if I recall correctly... long time ago :) and yes one of first few books published :) when books were worth writing :)

**Reply by theArianit** [reply]
I hope I can bother you with another question. Is fresh a SPA framework? Can it (island architecture maybe) be used as a SPA?

**Comment by OldManFroggy** [comment]
np, anytime, I spend way too much in Deno, and related techs at the moment
^^

just watch deno announcements in the next few months, I expect the npm/deno
issues may no longer be a thing, although to be honest using esm.sh for a
lot of things that are not native is not so bad; and in most cases porting
existing node/non-esm code to deno compliant via esm.sh or native deno is
not so bad these days -> Deno is production ready, Fresh is getting there,
and Deno Deploy is just a great DX (still needs maturity in operational
controls and observability, but will get there soon enough, and there are
ways to work around that.)

**Reply by rojvv** [reply]
you're not replying the comment btw

**Reply by theArianit** [reply]
I'll have to check how to work with node modules in deno

**Reply by OldManFroggy** [reply]
@roj1512 thanks,  ack used email reply, that will teach me

**Reply by OldManFroggy** [reply]
@theArianit in most cases you can get a import url by just going to your browser and going to the following url.

https://esm.sh/ + name of the nodejs import for example @headlessui/react

would advise reading information at the following [link](https://deno.land/manual/node/cdns#esmsh) and also deno documentation on imports.

you will run into some compatibility issues between react and preact, but in most cases an alias can fix that here is a [preact doc](https://preactjs.com/guide/v8/switching-to-preact/) on that aspect also.
